pitch.md

```diff
@@ -1,69 +1,172 @@
-# Project Title (replace me)
+# Compile-Time Programming
 
 ## Overview
-_What's this project about, briefly?_
 
+There are aspects of programming languages that operate during compile time.
+Examples are simple constant expressions, type checking, and macros. Less common
+are dependent types, correctness proofs (relatedly), and LISP-style abstract syntax tree macros.
+
+However, these are features which are extremely difficult to _add_ to an existing language.
+And they are nonetheless so desirable that people will go to extreme lengths to get them
+when they aren't available. (See: C++ template programming, or anything Andrei Alexandrescu writes)
+Furthermore, the nastiness of C++'s templates are not just inherited from C++'s
+general style. In Scala, similar things start to happen when you try to implement
+dependent types.
+
+Ultimately, the real problem is that programmers only have incidental control
+over compile-time behavior. I want to design a language where there is
+direct and encouraged access to compile-time behavior. This is extremely ambitious,
+so there are many limits that will be discussed later.
 
 ## The user and a language
-This section describes who the project would serve and why a language might be a
-good way to meet their needs.
 
 
 ### What's the need?
-_What need is met by your idea? Who are you helping? What is that person's
-experience like now? What would their experience be like if you could help 
-them?_
 
+The big need I see is that with current languages, extensibility ends at
+compile-time features. That is not completely true: again, it is _possible_
+to develop a lot of functionality by leveraging what already exists,
+especially when you get into the realm of macros.  
+And Haskell encourages modifications to the compiler, such that something
+like Idris (which has dependent types) can be built off of it, which is also
+pretty successful. Even so, these are not particularly friendly.
+I want to explore the possibility of extensibility in this region as well.
+
+There is also a second need.
+Supposedly, this would be a general-purpose programming language, whose compile-time features
+were especially easy to extend. That could mean that it would be extremely useful
+for people writing code that needs exotic compile-time features, such as verification:
+but, in actuality, if it were to be practically useful, it would probably be as a playground
+for working with different variations on compile-time concepts such as dependent types.
+So it is also potentially useful as a programming language researcher's aide.
```

Comment:
You seem pretty excited about dependent types, is there a project that you could do that would focus more on this but have a smaller scope? Maybe something for implementing dependent types in a language that doesn't have them?